tc/autotuner/autotuner.h

```diff
@@ -0,0 +1,214 @@
+  /// threads.
+  /// This function must be specialized per Backend
+  /// TODO: virtual inheritance?
+  void doEvaluate(size_t device, size_t populationSize, Printer& printer);
+
+  const MappingOptionsType& bestMappingOption() {
+    std::lock_guard<std::mutex> lock(bestTimeMtx_);
+    return bestMappingOptions_;
+  }
+
+ private:
+  static constexpr int kReducedWarmupIterations = 2;
+  static constexpr int kReducedBenchmarkIterations = 10;
+  static constexpr int kEarlyPruneFactor = 5;
+
+  /// Synchronization related objects
+  /// The main invariant is that we always try to compile and evaluate
+  /// exactly searchStrategy->population.size() candidates.
+  /// If a candidate fails compilation we still add a null Executor so that
+  /// the invariant holds.
+  /// This way it is easy to implement multi-threaded termination by just
+  /// taking an atomic counter and pushing/popping the queues under lock until
+  /// we have evaluated searchStrategy->population.size() compilation results.
+  std::mutex bestTimeMtx_;
+  std::mutex executorsMutex_;
+  std::atomic_bool stopRequested_;
+  std::atomic_size_t currentCompilationJob_;
+  std::atomic_size_t numEvaluations_;
+  std::queue<std::unique_ptr<typename Backend::ExecutorType>> executors_;
+  std::queue<CandidateConfiguration*> configurations_;
+
+  /// inputs
+  lang::TreeRef kTcTree_;
+  const MappingOptionsType kBaseMapping_;
+  const std::unordered_map<size_t, std::vector<const DLConstTensor*>> kInputs_;
+  std::unordered_map<size_t, std::vector<const DLTensor*>> outputs_;
+
+  // results
+  size_t bestTime_;
+  MappingOptionsType bestMappingOptions_;
+
+  // backing options cache
+  std::shared_ptr<OptionsCache<Backend>>& optionsCache_;
+};
+} // namespace detail
+
+/**
+ * An Autotuner provides the basic interface to run a SearchStrategy over a
+ * particular Backend.
+ * Under the hood it instantiates a TuningHarness<Backend> and a
+ * SearchStrategy. The Autotuner then calls TuningHarness<Backend>::run()
+ * on the SearchStrategy.
+ */
+template <typename Backend, typename SearchStrategy>
+class Autotuner {
+ public:
+  using BackendType = Backend;
+  using MappingOptionsType = typename BackendType::MappingOptionsType;
+  using OptionsCacheType = tc::autotune::OptionsCache<Backend>;
+
+  /// Parse and store the underlying tree representation for all the TC
+  /// functions provided.
+  explicit Autotuner(const std::string& tc);
+
+  /// Runs autotuning on the TC function tcEntryPoint.
+  /// This assumes input and output tensors of proper sizes have been
+  /// initialized on each GPU. For now this initialization is left to the
+  /// client of the Autotuner.
+  /// Tuning performs a search of template type SearchStrategy starting from
+  /// baseMapping.
+  ///
+  /// Alternatively an OptionsCache cacheFileName serialized path
+  /// can be specified from which the tuner recover multiple starting points
+  /// and preempt the baseMapping. This can be useful in a reinforcement
```

Comment:
do you mean ", preempting" instead of "and preempt"?
Does that mean that `baseMapping` gets ignored in that case?
Wouldn't it then make more sense to provide either the baseMapping or
a cacheFileName, but not both?

Reply:
impl concats them, fixing the doc